tools/crm_rule.c

```diff
@@ -0,0 +1,266 @@
+{
+    pe_working_set_t *data_set = NULL;
+    xmlNode *cib_constraints = NULL;
+    xmlNode *match = NULL;
+    xmlXPathObjectPtr xpathObj = NULL;
+    pe_eval_date_result_t result;
+    char *xpath = NULL;
+    int rc = pcmk_ok;
+    int max = 0;
+
+    /* Populate working set from CIB query */
+    if (inputXml == NULL) {
+        rc = cib_conn->cmds->query(cib_conn, NULL, &inputXml, cib_scope_local | cib_sync_call);
+        if (rc != pcmk_ok) {
+            return rc;
+        }
+    }
+
+    /* Populate the working set instance */
+    data_set = pe_new_working_set();
+    if (data_set == NULL) {
+        rc = -ENOMEM;
+        goto bail;
+    }
+
+    data_set->input = inputXml;
+    data_set->now = effective_date;
+
+    /* Unpack everything. */
+    cluster_status(data_set);
+
+    /* Rules are under the constraints node in the XML, so first find that. */
+    cib_constraints = get_object_root(XML_CIB_TAG_CONSTRAINTS, data_set->input);
+
+    /* Get all rules matching the given ID, which also have only a single date_expression
+     * child whose operation is not 'date_spec'.  This is fairly limited, but it's hard
+     * to check expressions more complicated than that.
+     */
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec']", rule_id);
+    xpathObj = xpath_search(cib_constraints, xpath);
+
+    max = numXpathResults(xpathObj);
+    if (max == 0) {
+        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        rc = 0;
+        goto bail;
+    } else if (max > 1) {
+        CMD_ERR("More than one date_expression in %s is not supported", rule_id);
+        rc = CRM_EX_ERROR;
+        goto bail;
+    }
+
+    match = getXpathResult(xpathObj, 0);
+
+    /* We should have ensured both of these pass with the xpath query above, but
+     * double checking can't hurt.
+     */
+    CRM_ASSERT(match != NULL);
+    CRM_ASSERT(find_expression_type(match) == time_expr);
+
+    result = pe_eval_date_expression(match, effective_date);
+
+    if (result == date_after_range) {
+        printf("Rule %s is expired\n", rule_id);
+        rc = 1;
+    } else if (result == date_within_range) {
+        printf("Rule %s is still in effect\n", rule_id);
+        rc = 0;
+    } else if (result == date_before_range) {
+        printf("Rule %s has not yet taken effect\n", rule_id);
+        rc = 1;
+    } else {
+        printf("Could not determine whether rule %s is expired\n", rule_id);
+        rc = 1;
+    }
+
+bail:
+    free(xpath);
+    freeXpathObject(xpathObj);
+    pe_free_working_set(data_set);
+    return rc;
+}
+
+int
+main(int argc, char **argv)
+{
+    cib_t *cib_conn = NULL;
+
+    int flag = 0;
+    int option_index = 0;
+
+    char *rule_id = NULL;
+    crm_time_t *rule_date = NULL;
+
+    xmlNode *input = NULL;
+    char *input_xml = NULL;
+    char *input_file = getenv("CIB_file");
+
+    int rc = pcmk_ok;
+    crm_exit_t exit_code = CRM_EX_OK;
```

Comment:
(background for later comments ...)

exit code vs rc:  functions can return anything, with a common model in pacemaker being an int set to the system errno or one of the custom pacemaker error codes. The exit status code a program returns to the OS, on the other hand, is greatly restricted, e.g. it must be in the 0-255 range, and only a subset of that is practically usable from within the program. We created crm_exit_t to represent that.